### PR TITLE
Add coverage for missing API (except executors)

### PR DESCRIPTION
**Pull request by brawner**
This adds coverage for most of the missing API that I've been able to identify, which were for the most part simple methods. There is still a largish chunk related to executors that is to come.

Signed-off-by: Stephen Brawner <brawner@gmail.com>

### CODE REVIEW

File: rclcpp/test/rclcpp/allocator/test_allocator_common.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#include <gtest/gtest.h>
+
+#include "rclcpp/allocator/allocator_common.hpp"
+
+TEST(TestAllocatorCommon, retyped_allocate) {
+  std::allocator<int> allocator;
+  void * untyped_allocator = &allocator;
+  void * allocated_mem =
+    rclcpp::allocator::retyped_allocate<std::allocator<char>>(1u, untyped_allocator);
+  EXPECT_NE(nullptr, allocated_mem);
```
Comment (clalancette) [inline rclcpp/test/rclcpp/allocator/test_allocator_common.cpp:24]: Should this be an `ASSERT_NE`?  That is, what happens during the `retyped_deallocate` below if this happens to be nullptr?

File: rclcpp/test/rclcpp/allocator/test_allocator_common.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        allocated_mem, untyped_allocator);
+    };
+  EXPECT_NO_THROW(code());
+
+  allocated_mem = allocator.allocate(1);
+  ASSERT_NE(nullptr, allocated_mem);
+  void * reallocated_mem =
+    rclcpp::allocator::retyped_reallocate<int, std::allocator<int>>(
+    allocated_mem, 2u, untyped_allocator);
+  EXPECT_NE(nullptr, reallocated_mem);
```
Comment (clalancette) [inline rclcpp/test/rclcpp/allocator/test_allocator_common.cpp:37]: Same question here as to whether it should be `ASSERT_NE`.

File: rclcpp/test/rclcpp/allocator/test_allocator_deleter.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  EXPECT_EQ(&allocator, deleter2.get_allocator());
+
+  rclcpp::allocator::AllocatorDeleter<std::allocator<int>> deleter3(deleter2);
+  EXPECT_EQ(&allocator, deleter3.get_allocator());
+}
+
+TEST(TestAllocatorDeleter, delete) {
+  std::allocator<int> allocator;
+  int * some_mem = allocator.allocate(1u);
+
```
Comment (clalancette) [inline rclcpp/test/rclcpp/allocator/test_allocator_deleter.cpp:39]: Maybe add an `ASSERT_NE(nullptr, some_mem);` ?

File: rclcpp/test/rclcpp/exceptions/test_exceptions.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  EXPECT_THROW(
+    rclcpp::exceptions::throw_from_rcl_error(RCL_RET_ERROR, ""),
+    rclcpp::exceptions::RCLError);
+
+  RCLCPP_EXPECT_THROW_EQ(
+    rclcpp::exceptions::throw_from_rcl_error(RCL_RET_OK, ""),
+    std::invalid_argument("ret is RCL_RET_OK"));
+
+  {
+    auto mock = mocking_utils::patch_and_return("lib:rclcpp", rcl_get_error_state, nullptr);
```
Comment (clalancette) [inline rclcpp/test/rclcpp/exceptions/test_exceptions.cpp:45]: I forget how the mocking works; do we need to "unmock" after this, or does leaving the scope do that?

File: rclcpp/test/rclcpp/test_rate.cpp [hunk 5]
```diff
@@ -98,3 +100,23 @@ TEST(TestRate, wall_rate_basics) {
   delta = five - four;
   EXPECT_GT(epsilon, delta);
 }
+
+TEST(TestRade, from_double) {
```
Comment (clalancette) [inline rclcpp/test/rclcpp/test_rate.cpp:104]:
```suggestion
TEST(TestRate, from_double) {
```

**Review by clalancette — CHANGES_REQUESTED** [review]
I've got a few small things to fix here.  Generally looks good, though.

**Comment by clalancette** [comment]
(oh, and cpplint seems to be very displeased)

File: rclcpp/test/rclcpp/test_timer.cpp [hunk 6]
```diff
@@ -191,3 +192,20 @@ TEST_F(TestTimer, test_bad_arguments) {
...
+    std::chrono::milliseconds(1),
+    [&timer_ptr](rclcpp::TimerBase & timer) {
+      timer_ptr = &timer;
+    });
+  auto start = std::chrono::steady_clock::now();
+  while (nullptr == timer_ptr &&
+    (std::chrono::steady_clock::now() - start) < std::chrono::milliseconds(100))
+  {
+    executor->spin_some();
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (hidmic) [inline rclcpp/test/rclcpp/test_timer.cpp:208]: @brawner nit: you could also use `spin_once(std::chrono::milliseconds(10))` and let it sleep.

File: rclcpp/test/rclcpp/allocator/test_allocator_common.cpp (hunk 1 above)
Comment (brawner) [reply rclcpp/test/rclcpp/allocator/test_allocator_common.cpp]: Good call. Done

File: rclcpp/test/rclcpp/allocator/test_allocator_common.cpp (hunk 2 above)
Comment (brawner) [reply rclcpp/test/rclcpp/allocator/test_allocator_common.cpp]: Done

File: rclcpp/test/rclcpp/allocator/test_allocator_deleter.cpp (hunk 3 above)
Comment (brawner) [reply rclcpp/test/rclcpp/allocator/test_allocator_deleter.cpp:42]: Done

File: rclcpp/test/rclcpp/allocator/test_allocator_deleter.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  rclcpp::allocator::set_allocator_for_deleter<AllocatorT, AllocatorT>(&deleter, &allocator2);
+  EXPECT_EQ(&allocator2, deleter.get_allocator());
+
+  auto throwing_statement = [&allocator]() {
+      DeleterT * null_del_ptr = nullptr;
+      rclcpp::allocator::set_allocator_for_deleter<AllocatorT, AllocatorT>(
+        null_del_ptr, &allocator);
+    };
+  RCLCPP_EXPECT_THROW_EQ(
+    throwing_statement(),
```
[the remark this answers is not in the document]
Comment (brawner) [reply rclcpp/test/rclcpp/allocator/test_allocator_deleter.cpp:63]: My compiler was having trouble parsing the templated function inside a macro.

File: rclcpp/test/rclcpp/exceptions/test_exceptions.cpp (hunk 4 above)
Comment (brawner) [reply rclcpp/test/rclcpp/exceptions/test_exceptions.cpp:45]: The scope block takes care of that, which is awesome

File: rclcpp/test/rclcpp/test_future_return_code.cpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+#include <gtest/gtest.h>
+
+#include <string>
+
+#include "rclcpp/future_return_code.hpp"
+
+TEST(TestFutureReturnCode, to_string) {
+  EXPECT_EQ(
+    std::string("Unknown enum value (-1)"), rclcpp::to_string(rclcpp::FutureReturnCode(-1)));
```
[the remark this answers is not in the document]
Comment (brawner) [reply rclcpp/test/rclcpp/test_future_return_code.cpp]: Done

File: rclcpp/test/rclcpp/test_parameter_client.cpp [hunk 9]
```diff
@@ -159,3 +161,24 @@ TEST_F(TestParameterClient, sync_parameter_event_subscription) {
...
+    std::runtime_error("Parameter 'not_a_parameter' is not set"));
+}
+
+/*
+   Coverage for async waiting/is_ready
+ */
+TEST_F(TestParameterClient, sync_parameter_is_ready) {
+  rclcpp::SyncParametersClient client(node);
+  EXPECT_TRUE(client.wait_for_service());
+  EXPECT_TRUE(client.service_is_ready());
```
[the remark this answers is not in the document]
Comment (brawner) [reply rclcpp/test/rclcpp/test_parameter_client.cpp:183]: What would be flakey about it? If `wait_for_service()` succeeds then `service_is_ready()` should also succeed?

File: rclcpp/test/rclcpp/test_rate.cpp (hunk 5 above)
Comment (brawner) [reply rclcpp/test/rclcpp/test_rate.cpp]: Done

File: rclcpp/test/rclcpp/test_timer.cpp (hunk 6 above)
Comment (brawner) [reply rclcpp/test/rclcpp/test_timer.cpp]: Switched to spin_once

**Comment by brawner** [comment]
Fixed up cpplint. That's what I get for forgetting to re-include the longer running linters when testing locally

Testing `--packages-select rclcpp`
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12438/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7405/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10148/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12351/)

**Review by clalancette — APPROVED** [review]
Looks good with green CI.

File: rclcpp/test/rclcpp/test_parameter_client.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply rclcpp/test/rclcpp/test_parameter_client.cpp:183]: Oof, I didn't grok the code clearly. Nevermind.

**Review by hidmic — APPROVED** [review]
LGTM too!

**Comment by brawner** [comment]
Windows is always different. Removing check on rcl_allocator.state in `test_allocator_common.cpp`

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12440/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7407/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10150/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12353/)

**Pull request merged**